src/Mvc/Mvc.Core/src/Infrastructure/ProblemDetailsFactory.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc.Infrastructure
+{
+    /// <summary>
+    /// Factory to produce <see cref="ProblemDetails" /> and <see cref="ValidationProblemDetails" />.
+    /// </summary>
+    public abstract class ProblemDetailsFactory
+    {
+        /// <summary>
+        /// Creates a <see cref="ProblemDetails" /> instance that configures defaults based on values specified in <see cref="ApiBehaviorOptions" />.
+        /// </summary>
+        /// <param name="httpContext">The <see cref="HttpContext" />.</param>
+        /// <param name="statusCode">The value for <see cref="ProblemDetails.Status"/>.</param>
+        /// <param name="title">The value for <see cref="ProblemDetails.Title" />.</param>
+        /// <param name="type">The value for <see cref="ProblemDetails.Type" />.</param>
+        /// <param name="detail">The value for <see cref="ProblemDetails.Detail" />.</param>
+        /// <param name="instance">The value for <see cref="ProblemDetails.Instance" />.</param>
+        /// <returns>The <see cref="ProblemDetails"/> instance.</returns>
+        public abstract ProblemDetails CreateProblemDetails(
+            HttpContext httpContext,
+            int? statusCode = null,
+            string title = null,
+            string type = null,
+            string detail = null,
+            string instance = null);
+
+        /// <summary>
+        /// Creates a <see cref="ValidationProblemDetails" /> instance that configures defaults based on values specified in <see cref="ApiBehaviorOptions" />.
+        /// </summary>
+        /// <param name="httpContext">The <see cref="HttpContext" />.</param>
+        /// <param name="modelStateDictionary">The <see cref="ModelStateDictionary" />.</param>
+        /// <param name="statusCode">The value for <see cref="ProblemDetails.Status"/>.</param>
+        /// <param name="title">The value for <see cref="ProblemDetails.Title" />.</param>
+        /// <param name="type">The value for <see cref="ProblemDetails.Type" />.</param>
+        /// <param name="detail">The value for <see cref="ProblemDetails.Detail" />.</param>
+        /// <param name="instance">The value for <see cref="ProblemDetails.Instance" />.</param>
+        /// <returns>The <see cref="ValidationProblemDetails"/> instance.</returns>
+        public abstract ValidationProblemDetails CreateValidationProblemDetails(
```

Comment:
Going back to an earlier question about what would you do if you needed to consistently return 422 for `ValidationProblem` - this would be the API to override this. This is called by `InvalidModelStateFilter` and `ControllerBase.ValidationProblem`